src/dmd/e2ir.d

```diff
@@ -3603,23 +3647,46 @@ elem *toElem(Expression e, IRState *irs)
                 }
             }
 
+            elem *eeq = null;
+            elem *ethis2 = null;
             elem *ethis;
             Symbol *sfunc = toSymbol(de.func);
             elem *ep;
-            if (de.func.isNested())
+            if (de.func.isNested() && !de.func.isThis())
             {
                 ep = el_ptr(sfunc);
                 if (de.e1.op == TOK.null_)
                     ethis = toElem(de.e1, irs);
                 else
-                    ethis = getEthis(de.loc, irs, de.func);
+                    ethis = getEthis(de.loc, irs, de.func, de.func.toParent4());
+                if (de.vthis2)
+                {
+                    elem *ectx0 = el_una(OPind, ethis.Ety, el_copytree(ethis2));
+                    elem *eeq1 = el_bin(OPeq, ethis.Ety, ectx0, ethis);
+                    ethis = el_copytree(ectx0);
+                    eeq = el_combine(eeq1, eeq);
+                }
             }
             else
             {
                 ethis = toElem(de.e1, irs);
                 if (de.e1.type.ty != Tclass && de.e1.type.ty != Tpointer)
                     ethis = addressElem(ethis, de.e1.type);
 
+                if (de.vthis2)
+                {
```

Comment:
you are mixing ast manipulationn and elem_tree manipulation, maybe the ast manipulation could be done in the frontend?

Reply:
I don't see a problem with it int this specific scenario since no semantic is run and the types are filled-in manually and since ast entities are the primary feed to the visitor. the reason I use ast nodes instead of elems is commented there `// using toSymbol directly because vthis2 may be a closure var` the problem is that `toSymbol` unlike `toElem` doesn't handle closure variables which `vthis2` often is.